source/fe/fe_dgq.cc

```diff
@@ -419,61 +419,51 @@ FE_DGQ<dim, spacedim>::get_prolongation_matrix(
   AssertIndexRange(child, GeometryInfo<dim>::n_children(refinement_case));
 
   // initialization upon first request
-  if (this->prolongation[refinement_case - 1][child].n() == 0)
-    {
-      std::lock_guard<std::mutex> lock(this->mutex);
-
-      // if matrix got updated while waiting for the lock
-      if (this->prolongation[refinement_case - 1][child].n() ==
-          this->n_dofs_per_cell())
-        return this->prolongation[refinement_case - 1][child];
-
-      // now do the work. need to get a non-const version of data in order to
-      // be able to modify them inside a const function
-      FE_DGQ<dim, spacedim> &this_nonconst =
-        const_cast<FE_DGQ<dim, spacedim> &>(*this);
-      if (refinement_case == RefinementCase<dim>::isotropic_refinement)
-        {
-          std::vector<std::vector<FullMatrix<double>>> isotropic_matrices(
-            RefinementCase<dim>::isotropic_refinement);
-          isotropic_matrices.back().resize(
-            GeometryInfo<dim>::n_children(RefinementCase<dim>(refinement_case)),
-            FullMatrix<double>(this->n_dofs_per_cell(),
-                               this->n_dofs_per_cell()));
-          if (dim == spacedim)
+  [[maybe_unused]] static bool once = [&]() {
```

Comment:
Static variables in member functions are only initialized once per set of template parameters.
I don't think this is what is intended here.

Reply:
```
#include <iostream>

template <int bar>
struct Test
{
public:
  void test() {
    static bool initialized = [&]{
      std::cout << "Initialize " << bar << std::endl;
      return true;
    }();
    std::cout << "Called test" << std::endl;
  }

private:
  int foo;
};

int main()
{
  Test<1> test;
  test.test();

  Test<1> test2;
  test2.test();
}
```
Produces:
```
Initialize 1
Called test
Called test
```

Reply:
Indeed, we need to be able to call the function for different polynomial degrees, that get used in the same program. For example via Taylor-Hood elements (pressure degree `p`, velocity degree `p+1`), most more complex programs with different fields, or all `hp` applications that use solution transfer and similar mechanisms.

I also would be happy if we could keep the of static variables and other data structures that live outside `main()` small, because of the associated costs. I think we were worse a few years ago, but I still see a substantial number of operations done in `call_init` below main, e.g. things from `multithread_info` or the static `MappingQ1`.

Reply:
> Static variables in member functions are only initialized once per set of template parameters.
I don't think this is what is intended here.

Ah, missed that.